crates/bevy_audio/src/audio_input.rs

```diff
@@ -0,0 +1,180 @@
+/// the underlying stream and event system are also dropped.
+pub struct AudioInputStream {
+    config: StreamConfig,
+    stream: Stream,
+    receiver: Receiver<AudioInputEvent>,
+}
+
+/// A frame of samples from an audio input source.
+#[derive(Event, Clone)]
+pub struct AudioInputEvent {
+    /// Interleaved samples recorded during this frame of audio input. Please check [`config`](`Self::config`)
+    /// for channel information and sample rate.
+    pub samples: Vec<f32>,
+    /// Information about this sample frame as provided by [`cpal`].
+    pub info: InputCallbackInfo,
+    /// Configuration used for the input device which recorded this sample frame.
+    pub config: StreamConfig,
+}
+
+impl AudioInputStream {
+    /// Try to create a default audio input stream. Returns [`None`] if a default input stream
+    /// cannot be created.
+    pub fn try_default() -> Option<Self> {
+        let host = cpal::default_host();
+        let input = host.default_input_device()?;
+        let config = input.default_input_config().ok()?.config();
+
+        trace!("Got Audio Input Device: {:?}: {:?}", input.name(), config);
+
+        let (tx, receiver) = channel();
+
+        // Create a clone of the configuration for the callback
+        let config_clone = config.clone();
+
+        let stream = input
+            .build_input_stream(
+                &config,
+                move |data: &[f32], info: &InputCallbackInfo| {
+                    let event = AudioInputEvent {
+                        samples: data.to_vec(),
+                        info: info.clone(),
+                        config: config_clone.clone(),
+                    };
+
+                    if let Err(error) = tx.send(event) {
+                        warn!("Audio Input Error: {}", error);
+                    };
+                },
+                move |error| {
+                    warn!("Audio Input Error: {}", error);
+                },
+                None,
+            )
+            .ok()?;
+
+        Some(Self {
+            config,
+            stream,
+            receiver,
+        })
+    }
+
+    /// Returns an [`Iterator`] over newly received [`AudioInputEvent`]'s. This does not block
+    /// if new events aren't available. This clears current events.
+    pub fn iter(&self) -> impl Iterator<Item = AudioInputEvent> + '_ {
+        self.receiver.try_iter()
+    }
+
+    /// Returns the [configuration information](`StreamConfig`) of this [`AudioInputStream`].
+    pub fn config(&self) -> &StreamConfig {
+        &self.config
+    }
+
+    /// Provides access to the underlying [`Stream`].
+    pub fn stream(&self) -> &Stream {
+        &self.stream
+    }
+
+    /// Start recording audio from this input.
+    pub fn start(&self) -> Result<(), PlayStreamError> {
+        self.stream.play()
+    }
+
+    /// Stop recording audio from this input.
+    pub fn stop(&self) -> Result<(), PauseStreamError> {
+        self.stream.pause()
+    }
```

Comment:
We should probably wrap the error type, to prevent "leaking the internals" and allow us to change the underlying implementation without introducing breaking changes

Reply:
Agreed!